### PR TITLE
Hari/groth16

### LINKED ISSUE #289: Migrate to Groth 16
- The verifier.sol for Groth16 can be created by extracting from `export-verifier` in zokrates
- The contract must have no hard coded verification key. We will need to register this at the time of deploying contracts. See `circuit-setup.mjs` in `nightfall-deployer`
- There were gas efficiency changes made to the current verifier.sol which was also a fork of the export-verifier from zokrates for GM17. Please see if these can be redone here too
- `nightfall-deployer`, `nightfall-client` and `nightfall-optimist` should move to Groth 16 when calling zokrates

### PR DESCRIPTION
**Pull request by dwebchapey**
fixes #289 

- Change Zokrates worker and env variables to `g16 , bellman`
- Modify vk structure to g16 format
- `Verifier.sol` and `Pairing.sol` files  from zokrates `export-verifier`

### CODE REVIEW

**Review by IlyasRidhuan — APPROVED** [review]

**Review by ChaitanyaKonda — APPROVED** [review]

File: nightfall-deployer/contracts/Verifier.sol [hunk 1]
```diff
@@ -59,31 +58,31 @@ library Verifier {
...
       proof.A = Pairing.G1Point(_proof[0], _proof[1]);
       proof.B = Pairing.G2Point([_proof[2], _proof[3]], [_proof[4], _proof[5]]);
       proof.C = Pairing.G1Point(_proof[6], _proof[7]);
 
-      vk.H = Pairing.G2Point([_vk[0],_vk[1]],[_vk[2],_vk[3]]);
-      vk.Galpha = Pairing.G1Point(_vk[4],_vk[5]);
-      vk.Hbeta = Pairing.G2Point([_vk[6],_vk[7]],[_vk[8],_vk[9]]);
-      vk.Ggamma = Pairing.G1Point(_vk[10],_vk[11]);
-      vk.Hgamma = Pairing.G2Point([_vk[12],_vk[13]],[_vk[14],_vk[15]]);
+      vk.alpha = Pairing.G1Point(_vk[4],_vk[5]);
```
Comment (MirandaWood) [inline nightfall-deployer/contracts/Verifier.sol:71]: Wondering if there is any reason we don't start from 0 here? Seems like we dont use elts 0 to 3 of the vk

File: nightfall-deployer/contracts/Verifier.sol [hunk 2]
```diff
@@ -59,31 +58,31 @@ library Verifier {
...
+      vk.beta = Pairing.G2Point([_vk[6],_vk[7]],[_vk[8],_vk[9]]);
+      vk.gamma = Pairing.G2Point([_vk[10],_vk[11]],[_vk[12],_vk[13]]);
+      vk.delta = Pairing.G2Point([_vk[14],_vk[15]],[_vk[16],_vk[17]]);
```
Comment (MirandaWood) [inline nightfall-deployer/contracts/Verifier.sol:74]:
```suggestion
      vk.alpha = Pairing.G1Point(_vk[0],_vk[1]);
      vk.beta = Pairing.G2Point([_vk[2],_vk[3]],[_vk[4],_vk[5]]);
      vk.gamma = Pairing.G2Point([_vk[6],_vk[7]],[_vk[8],_vk[9]]);
      vk.delta = Pairing.G2Point([_vk[10],_vk[11]],[_vk[12],_vk[13]]);
```
This and the below got it working on my end :)

File: nightfall-deployer/contracts/Verifier.sol [hunk 3]
```diff
@@ -59,31 +58,31 @@ library Verifier {
...
 
-      // vk.query.length = (_vk.length - 16)/2;
+      // vk.gamma_abc.length = (_vk.length - 18)/2;
       uint j = 0;
-      for (uint i = 16; i < _vk.length; i+=2) {
-        vk.query[j++] = Pairing.G1Point(_vk[i], _vk[i+1]);
+      for (uint i = 18; i < _vk.length; i+=2) {
+        vk.gamma_abc[j++] = Pairing.G1Point(_vk[i], _vk[i+1]);
       }
```
Comment (MirandaWood) [inline nightfall-deployer/contracts/Verifier.sol:80]:
```suggestion
      if (_vk.length > 14) {
        for (uint i = 14; i < _vk.length; i+=2) {
        vk.gamma_abc[(i-14)/2] = Pairing.G1Point(
            _vk[i], _vk[i+1]
        );
        }
      }
```
(I think using `j` will also work here, the main thing is the bounds)

**Review by MirandaWood — CHANGES_REQUESTED** [review]

File: nightfall-deployer/contracts/Verifier.sol (hunk 2 above)
Comment (dwebchapey) [reply nightfall-deployer/contracts/Verifier.sol:74]: Yes . It was a copy mistake from my side . Thanks for noting it @MirandaWood

**Review by MirandaWood — APPROVED** [review]

**Pull request merged**